From: Dave Airlie <airlied@redhat.com>
Date: Fri Mar 28 09:18:45 2014 +1000
Subject: drm/ast: add AST 2400 support.
Patch-Mainline: Queued in subsystem maintainer repository
Git-commit: 4db5b603834b7ca490c53cbb26dbf9dfe4c83c17
Git-repo: git://people.freedesktop.org/~airlied/linux/
References: bnc#880007
Signed-off-by: Egbert Eich <eich@suse.com>

This is ported from the userspace driver.

Untested on any ast2400 hw so far.

Signed-off-by: Dave Airlie <airlied@redhat.com>
---
 drivers/gpu/drm/ast/ast_drv.h  |    1 +
 drivers/gpu/drm/ast/ast_main.c |    8 ++++++--
 drivers/gpu/drm/ast/ast_mode.c |    2 +-
 drivers/gpu/drm/ast/ast_post.c |    6 +++---
 4 files changed, 11 insertions(+), 6 deletions(-)

--- linux-3.0-SLE11-SP3.orig/drivers/gpu/drm/ast/ast_drv.h
+++ linux-3.0-SLE11-SP3/drivers/gpu/drm/ast/ast_drv.h
@@ -61,6 +61,7 @@ enum ast_chip {
 	AST2200,
 	AST2150,
 	AST2300,
+	AST2400,
 	AST1180,
 };
 
--- linux-3.0-SLE11-SP3.orig/drivers/gpu/drm/ast/ast_main.c
+++ linux-3.0-SLE11-SP3/drivers/gpu/drm/ast/ast_main.c
@@ -71,7 +71,10 @@ static int ast_detect_chip(struct drm_de
 		ast->chip = AST1100;
 		DRM_INFO("AST 1180 detected\n");
 	} else {
-		if (dev->pdev->revision >= 0x20) {
+		if (dev->pdev->revision >= 0x30) {
+			ast->chip = AST2400;
+			DRM_INFO("AST 2400 detected\n");
+		} else if (dev->pdev->revision >= 0x20) {
 			ast->chip = AST2300;
 			DRM_INFO("AST 2300 detected\n");
 		} else if (dev->pdev->revision >= 0x10) {
@@ -129,7 +132,7 @@ static int ast_get_dram_info(struct drm_
 	else
 		ast->dram_bus_width = 32;
 
-	if (ast->chip == AST2300) {
+	if (ast->chip == AST2300 || ast->chip == AST2400) {
 		switch (data & 0x03) {
 		case 0:
 			ast->dram_type = AST_DRAM_512Mx16;
@@ -371,6 +374,7 @@ int ast_driver_load(struct drm_device *d
 	if (ast->chip == AST2100 ||
 	    ast->chip == AST2200 ||
 	    ast->chip == AST2300 ||
+	    ast->chip == AST2400 ||
 	    ast->chip == AST1180) {
 		dev->mode_config.max_width = 1920;
 		dev->mode_config.max_height = 2048;
--- linux-3.0-SLE11-SP3.orig/drivers/gpu/drm/ast/ast_mode.c
+++ linux-3.0-SLE11-SP3/drivers/gpu/drm/ast/ast_mode.c
@@ -389,7 +389,7 @@ static void ast_set_ext_reg(struct drm_c
 	ast_set_index_reg_mask(ast, AST_IO_CRTC_PORT, 0xa8, 0xfd, jregA8);
 
 	/* Set Threshold */
-	if (ast->chip == AST2300) {
+	if (ast->chip == AST2300 || ast->chip == AST2400) {
 		ast_set_index_reg(ast, AST_IO_CRTC_PORT, 0xa7, 0x78);
 		ast_set_index_reg(ast, AST_IO_CRTC_PORT, 0xa6, 0x60);
 	} else if (ast->chip == AST2100 ||
--- linux-3.0-SLE11-SP3.orig/drivers/gpu/drm/ast/ast_post.c
+++ linux-3.0-SLE11-SP3/drivers/gpu/drm/ast/ast_post.c
@@ -78,7 +78,7 @@ ast_set_def_ext_reg(struct drm_device *d
 	for (i = 0x81; i <= 0x8f; i++)
 		ast_set_index_reg(ast, AST_IO_CRTC_PORT, i, 0x00);
 
-	if (ast->chip == AST2300) {
+	if (ast->chip == AST2300 || ast->chip == AST2400) {
 		if (dev->pdev->revision >= 0x20)
 			ext_reg_info = extreginfo_ast2300;
 		else
@@ -102,7 +102,7 @@ ast_set_def_ext_reg(struct drm_device *d
 
 	/* Enable RAMDAC for A1 */
 	reg = 0x04;
-	if (ast->chip == AST2300)
+	if (ast->chip == AST2300 || ast->chip == AST2400)
 		reg |= 0x20;
 	ast_set_index_reg_mask(ast, AST_IO_CRTC_PORT, 0xb6, 0xff, reg);
 }
@@ -365,7 +365,7 @@ void ast_post_gpu(struct drm_device *dev
 	ast_open_key(ast);
 	ast_set_def_ext_reg(dev);
 
-	if (ast->chip == AST2300)
+	if (ast->chip == AST2300 || ast->chip == AST2400)
 		ast_init_dram_2300(dev);
 	else
 		ast_init_dram_reg(dev);
